File: examples/typescript/src/patterns/workflow/entitystore/user-profile-flow.ts
/*
 * Copyright (c) 2022-2026 Super Durable, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import {
  Attribute,
  StepList,
  booleanCodec,
  doubleCodec,
  int64Codec,
  rpc,
  stringCodec,
  type Context,
  type Flow,
  type PersistenceSchema,
  type RPCResult,
} from "@superdurable/dex";

import {
  dateCodec,
  metadataCodec,
  userProfileCodec,
  type UserProfile,
} from "./user-profile.js";

export const ENTITY_STORE_NAME = "entityStore";

export class UserProfileFlow implements Flow<void> {
  public readonly displayName = new Attribute("display_name", stringCodec)
    .syncToAttributeStore();
  public readonly email = new Attribute("email", stringCodec).syncToAttributeStore();
  public readonly marketingOptIn = new Attribute("marketing_opt_in", booleanCodec)
    .syncToAttributeStore();
  public readonly credits = new Attribute("credits", int64Codec).syncToAttributeStore();
  public readonly weight = new Attribute("weight", doubleCodec).syncToAttributeStore();
  public readonly lastLoggedInTime = new Attribute("last_logged_in_time", dateCodec)
    .syncToAttributeStore();
  public readonly metadata = new Attribute("metadata", metadataCodec)
    .syncToAttributeStore();

  public getFlowType(): string {
    return "UserProfileFlow";
  }

  public getSteps() {
    return StepList.empty();
  }

  public getPersistenceSchema(): PersistenceSchema {
    return {
      attributes: [
        this.displayName,
        this.email,
        this.marketingOptIn,
        this.credits,
        this.weight,
        this.lastLoggedInTime,
        this.metadata,
      ],
    };
  }

  @rpc({ inputCodec: userProfileCodec })
  public updateProfile(context: Context, profile: UserProfile): void {
    this.displayName.set(context, profile.displayName);
    this.email.set(context, profile.email);
    this.marketingOptIn.set(context, profile.marketingOptIn);
    this.credits.set(context, BigInt(profile.credits));
    this.weight.set(context, profile.weight);
    this.lastLoggedInTime.set(context, profile.lastLoggedInTime);
    this.metadata.set(context, profile.metadata);
  }

  @rpc({ outputCodec: userProfileCodec })
  public getProfile(context: Context): RPCResult<UserProfile> {
    return {
      output: {
        displayName: this.displayName.get(context),
        email: this.email.get(context),
        marketingOptIn: this.marketingOptIn.get(context),
        credits: Number(this.credits.get(context)),
        weight: this.weight.get(context),
        lastLoggedInTime: this.lastLoggedInTime.get(context),
        metadata: this.metadata.get(context),
      },
    };
  }

  @rpc()
  public clearProfile(context: Context): void {
    this.displayName.delete(context);
    this.email.delete(context);
    this.marketingOptIn.delete(context);
    this.credits.delete(context);
    this.weight.delete(context);
    this.lastLoggedInTime.delete(context);
    this.metadata.delete(context);
  }
}

export const userProfileFlow = new UserProfileFlow();

File: examples/java/src/main/java/io/superdurable/dex/patterns/workflow/entitystore/UserProfileFlow.java
/*
 * Copyright (c) 2022-2026 Super Durable, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package io.superdurable.dex.patterns.workflow.entitystore;

import io.superdurable.dex.Attribute;
import io.superdurable.dex.Context;
import io.superdurable.dex.Flow;
import io.superdurable.dex.PersistenceSchema;
import io.superdurable.dex.RPC;
import io.superdurable.dex.RPCResult;
import io.superdurable.dex.StepList;
import org.springframework.stereotype.Component;

/** Keeps one user profile in Dex and projects its Attributes to PostgreSQL. */
@Component
public class UserProfileFlow implements Flow<Void> {
    public static final String STORE_NAME = "entityStore";

    public final Attribute<String> displayName =
            Attribute.define("display_name", String.class).syncToAttributeStore();
    public final Attribute<String> email =
            Attribute.define("email", String.class).syncToAttributeStore();
    public final Attribute<Boolean> marketingOptIn =
            Attribute.define("marketing_opt_in", Boolean.class).syncToAttributeStore();

    @Override
    public StepList<Void> getSteps() {
        return StepList.empty();
    }

    @Override
    public PersistenceSchema getPersistenceSchema() {
        return PersistenceSchema.of(displayName, email, marketingOptIn);
    }

    @RPC
    public void updateProfile(final Context context, final UserProfile profile) {
        profile.validate();
        displayName.set(context, profile.displayName);
        email.set(context, profile.email);
        marketingOptIn.set(context, profile.marketingOptIn);
    }

    @RPC
    public RPCResult<UserProfile> getProfile(final Context context) {
        return RPCResult.of(new UserProfile(
                displayName.get(context),
                email.get(context),
                marketingOptIn.get(context)));
    }

    @RPC
    public void clearProfile(final Context context) {
        displayName.delete(context);
        email.delete(context);
        marketingOptIn.delete(context);
    }
}
